File: homy/integrations/sonarr/widgets/queue/queue.py
"""*arr queue widget — Sonarr / Radarr / Lidarr / Readarr."""
import logging

from flask import jsonify, request

from homy.cache import cache_key, should_bypass_cache, widget_cache
from homy.integration_widget_util import get_integration_for_widget, get_widget_config
from homy import integration_widget_fetch as iwf

logger = logging.getLogger(__name__)

WIDGET = {
    'type': 'arr_queue',
    'name': '*arr Queue',
    'integration_types': ['sonarr', 'radarr', 'lidarr', 'readarr'],
    'default_size_x': 8,
    'default_size_y': 5,
    'icon': 'list-ordered',
    'config_schema': {
        'integration_id': {'type': 'text', 'label': 'Integration ID', 'default': ''},
        'max_items': {
            'type': 'select',
            'label': 'Max. Einträge',
            'options': ['5', '10', '15', '20'],
            'default': '10',
        },
        'show_progress': {
            'type': 'select',
            'label': 'Fortschritt-Balken',
            'options': ['Ja', 'Nein'],
            'default': 'Ja',
        },
        'show_eta': {
            'type': 'select',
            'label': 'ETA anzeigen',
            'options': ['Ja', 'Nein'],
            'default': 'Ja',
        },
        'show_protocol': {
            'type': 'select',
            'label': 'Protokoll-Badge (torrent/usenet)',
            'options': ['Ja', 'Nein'],
            'default': 'Nein',
        },
    },
}

CACHE_NS = 'arr_queue'
CACHE_TTL = 45


def register(app):
    @app.route('/api/arr_queue/data', methods=['GET'], endpoint='widget_arr_queue_data')
    def arr_queue_data():
        widget_id = request.args.get('widget_id')
        if not widget_id:
            return jsonify({'error': 'Bad Request', 'message': 'widget_id required'}), 400

        wcfg = get_widget_config(widget_id)
        if wcfg is None:
            return jsonify({'error': 'Not Found'}), 404

        integration, config, err, code = get_integration_for_widget(widget_id, ['sonarr', 'radarr', 'lidarr', 'readarr'])
        if err:
            return jsonify({'configured': False, 'online': False, 'message': err}), code

        ck = cache_key(CACHE_NS, widget_id)
        if not should_bypass_cache():
            cached = widget_cache.get(ck)
            if cached is not None:
                return jsonify(cached)

        try:
            result = iwf.fetch_arr_queue_list(integration, config, wcfg.get('max_items', 10))
            result['show_progress'] = wcfg.get('show_progress', 'Ja') == 'Ja'
            result['show_eta'] = wcfg.get('show_eta', 'Ja') == 'Ja'
            result['show_protocol'] = wcfg.get('show_protocol', 'Nein') == 'Ja'
            result['configured'] = True
            result['online'] = True
            widget_cache.set(ck, result, ttl=CACHE_TTL)
            return jsonify(result)
        except Exception as exc:
            logger.warning('arr_queue failed: %s', exc)
            return jsonify({'configured': True, 'online': False, 'message': str(exc)}), 200

File: homy/integrations/sonarr/widgets/queue/queue.js
window.WidgetRegistry.register('arr_queue', {
    _fmtBytes(b) {
        if (!b) return '';
        if (b >= 1073741824) return (b / 1073741824).toFixed(1) + ' GB';
        if (b >= 1048576) return (b / 1048576).toFixed(0) + ' MB';
        return (b / 1024).toFixed(0) + ' KB';
    },

    async render(container, widgetData, config) {
        container.innerHTML = '<div class="widget-loading"><div class="spinner"></div></div>';
        try {
            const data = await API.request(`/api/arr_queue/data?widget_id=${widgetData.id}`);
            container.innerHTML = '';
            if (!data.online) {
                container.innerHTML = `<div class="widget-error text-center" style="padding:12px;">${data.message || 'Not available'}</div>`;
                return;
            }
            const items = data.items || [];
            const statsHtml = `<div class="aq-stats">
                <div class="aq-stat"><span class="aq-val">${data.app_name || 'Queue'}</span><span class="aq-label">App</span></div>
                <div class="aq-stat"><span class="aq-val">${data.queue_count ?? 0}</span><span class="aq-label">Aktiv</span></div>
                <div class="aq-stat"><span class="aq-val">${data.queue_total ?? 0}</span><span class="aq-label">Gesamt</span></div>
            </div>`;

            const itemsHtml = items.length
                ? items.map(i => {
                    const pct = Math.min(100, Math.max(0, i.progress ?? 0));
                    const progressHtml = data.show_progress
                        ? `<div class="aq-progress-track"><div class="aq-progress-fill" style="width:${pct}%"></div></div>`
                        : '';
                    const proto = data.show_protocol && i.protocol
                        ? `<span class="aq-proto aq-proto-${i.protocol}">${i.protocol}</span>` : '';
                    const eta = data.show_eta && i.eta ? `<span class="aq-eta">${i.eta}</span>` : '';
                    const left = i.sizeleft ? this._fmtBytes(i.sizeleft) : '';
                    return `<div class="aq-item">
                        <div class="aq-item-header">
                            <span class="aq-item-title" title="${i.title}">${i.title}</span>
                            <div class="aq-item-badges">${proto}${eta}</div>
                        </div>
                        <div class="aq-item-meta">
                            <span>${i.status || ''}</span>
                            ${left ? `<span class="aq-left">${left} übrig</span>` : ''}
                            <span class="aq-pct">${pct}%</span>
                        </div>
                        ${progressHtml}
                    </div>`;
                }).join('')
                : '<div class="aq-empty">Queue leer</div>';

            container.innerHTML = `<div class="aq-container">${statsHtml}<div class="aq-list">${itemsHtml}</div></div>`;
        } catch (err) {
            container.innerHTML = `<div class="widget-error">${err.message}</div>`;
        }
    },
});
